drivers/i2s/i2s_ll_stm32.c

```diff
@@ -336,10 +350,21 @@ static int i2s_stm32_trigger(const struct device *dev, enum i2s_dir dir,
 			return -EIO;
 		}
 		irq_unlock(key);
-		stream->stream_disable(stream, dev);
-		stream->queue_drop(stream);
-		stream->state = I2S_STATE_READY;
-		stream->last_block = true;
+do_trigger_stop:
+		if (stream->state == I2S_STATE_RUNNING) {
+			stream->state = I2S_STATE_STOPPING;
+			/* indicate that the transition to I2S_STATE_STOPPING*/
+			/* is triggered by STOP command                      */
+			stream->tx_stop_for_drain = false;
+		} else if ((queue_is_empty(&stream->mem_block_queue) == false) ||
+					(ll_func_i2s_dma_busy(cfg->i2s))) {
+			stream->stream_disable(stream, dev);
+			stream->queue_drop(stream);
+			stream->state = I2S_STATE_READY;
+			stream->last_block = true;
```

Comment:
This doesn't comply with the API description which states:
```
   Stop the transmission / reception of data at the end of the current
   memory block. This trigger can be used in RUNNING state only and at
   first changes the interface state to STOPPING. When the current TX /
   RX block is transmitted / received the state is changed to READY.
   Subsequent START trigger will resume transmission / reception where
   it stopped.
```

I think the whole `if`/`else if`/`else` can be replaced with something like this: (also fixes the logic bug above)
```c
if (/* dma busy */) {
  stream->state = STOPPING
  stream->tx_stop_for_drain = false;
} else {
  stream->stream_disable(/* ... */); //i.e., stop the DMA
  //update state, switch to READY
}
```

_Make sure **not** to call `stream->queue_drop()`._